Persist referral link

stackseeder opened the pull request:
# Summary

Fixes #850 

When we use the referral link, referrer address is persisted in the state.

No visible change here. (No screenshot)

  # To Test

1. Copy referral link (or emulate any referral link like http(s)://<domain>/swap?referral=<address>)
- [] Paste url in another tab.
- [] Referrer address will be stored in the local storage like {"referrerAddress":"<Address>"}

ramirotw commented on src/components/Web3ReactManager/index.tsx:
```diff
@@ -48,6 +53,20 @@ export default function Web3ReactManager({ children }: { children: JSX.Element }
+      generateAffiliateLink((link: string) => dispatch(updateAffiliateLink({ affiliateLink: link })))
+    })
+  }
+
```
I'm not sure why this handler is needed to detect an account change. Why not just use the `account` property of `useWalletInfo` in a useEffect?

ramirotw commented on src/hooks/web3.ts:
```diff
@@ -57,17 +69,25 @@ export function useInactiveListener(suppress = false) {
+            })
+            .finally(() =>
+              generateAffiliateLink((link: string) => dispatch(updateAffiliateLink({ affiliateLink: link })))
+            )
```
Is this necessary? Remember that this file is not under the `custom` folder.

stackseeder replied:
@ramirotw do you want to move this file to custom folder?
It is used in too many files already.

ramirotw commented on src/utils/affiliateLinkUtil.ts:
```diff
@@ -0,0 +1,13 @@
+  if (accounts.length > 0) link = `${window.location.href}/#/?referral=${accounts[0]}`
+  console.log('GEN LINK =>', link)
+  if (callback) return callback(link)
+}
```
There's already a `useReferralLink` hook to do this.

ramirotw commented on package.json:
```diff
@@ -168,6 +168,7 @@
     "@gnosis.pm/cow-runner-game": "^0.2.9",
     "@gnosis.pm/dex-js": "^0.11.0",
     "@gnosis.pm/gp-v2-contracts": "^0.0.1-alpha.20",
+    "@metamask/detect-provider": "^1.2.0",
```
is this library really needed?

ramirotw reviewed (commented):
Check the build also which has a few errors https://github.com/gnosis/cowswap/pull/1025/checks?check_run_id=3135399538

ramirotw commented on yarn.lock:
```diff
@@ -2300,6 +2300,11 @@
+  version "1.2.0"
+  resolved "https://registry.yarnpkg.com/@metamask/detect-provider/-/detect-provider-1.2.0.tgz#3667a7531f2a682e3c3a43eaf3a1958bdb42a696"
+  integrity sha512-ocA76vt+8D0thgXZ7LxFPyqw3H7988qblgzddTDA6B8a/yU0uKV42QR/DhA+Jh11rJjxW0jKvwb5htA6krNZDQ==
+
```
This dependency can be removed

stackseeder replied:
I've removed it.

ramirotw commented on src/hooks/web3.ts:
```diff
@@ -49,25 +62,39 @@ export function useEagerConnect() {
+              if (referralLink) {
+                dispatch(updateAffiliateLink({ affiliateLink: referralLink }))
+              }
+            })
```
I still don't understand why do we need this finally clauses. Can you elaborate on why this file needs to be updated?

stackseeder replied:
I did not use useEffect before so we required this code. But now I am using useEffect, so don't need to specify the code in this file. I've removed it.

ramirotw commented on src/custom/components/Web3ReactManager/index.tsx:
```diff
@@ -48,6 +55,12 @@ export default function Web3ReactManager({ children }: { children: JSX.Element }
+      dispatch(updateAffiliateLink({ affiliateLink: referralLink }))
+    }
+  }, [account, referralLink, dispatch])
+
```
We shouldn't be updating either moving this Web3ReactManager file. Create a new hook which does this integration and put it in the App/index.ts file so it runs when the app renders.

alfetopito reviewed (changes requested):
I'm sorry Ivica, but that's now how our Modding system works.

Please do not move the original files out of their places, not modify the original files unless strictly necessary.
Take a look at this doc for reference https://docs.google.com/document/d/1-4b_BXevDPq1-anyx9qFnocshlUqom5W1lYuA2Y65us/edit#heading=h.lb282ue6cmsr
I'm happy to go over it with you if you have any questions.

ramirotw commented on src/hooks/web3.ts:
```diff
@@ -82,5 +86,5 @@ export function useInactiveListener(suppress = false) {
     }
     return undefined
-  }, [active, error, suppress, activate])
+  }, [active, error, suppress, activate, dispatch])
```
Changes in this file are outdated

ramirotw commented on src/state/affiliate/reducer.ts:
```diff
@@ -0,0 +1,17 @@
+
+export default createReducer(initialState, (builder) =>
+  builder.addCase(updateAffiliateLink, (state, action) => {
+    console.log('Triggered Updated Link')
```
nit: Remove this console.log

ramirotw reviewed (changes requested):
Looking good!

Just a few things to update

ramirotw commented on src/custom/state/index.ts:
```diff
@@ -19,7 +19,7 @@ import orders from './orders/reducer'
 import gas from 'state/gas/reducer'
 import { updateVersion } from 'state/global/actions'
-import affiliate from '@src/state/affiliate/reducer'
+import affiliate from '@src/custom/state/affiliate/reducer'
```
remove the `@src/custom/` path prefix

ramirotw reviewed (changes requested):
Please remove the remaining changes in the `src/hooks/web3.ts` file

alfetopito commented on src/custom/state/affiliate/updater.tsx:
```diff
@@ -0,0 +1,20 @@
+
+export default function ReferralLinkUpdater() {
+  const dispatch = useAppDispatch()
+  const referralLink = useReferralLink()
```
Actually, this returns your own affiliate link, the one you want to share with other people.
What we want here is to store what's coming from the url param when someone else opens your URL affiliate link.

You'll want to get the value Ramiro is fetching on this file from the PR https://github.com/gnosis/cowswap/pull/953/files#diff-af06422a6866f6f150604ce5491f449819b0882344b8da8b2387f6288fe74949L1

See the comments there.
The one where you said useless code? That's where you'll trigger your stuff :)

ramirotw replied:
> What we want here is to store what's coming from the url param when someone else opens your URL affiliate link.

@alfetopito I believe there's a misunderstanding with the ticket's description:

> Persist in the global state of the app, associated with the connected address, remember the user can change addresses any time

The ticket mentions the connected address while your comment says its the address from the url.

alfetopito reviewed (changes requested):
Great job, you the spirit of the mod files now :)

Comments below

alfetopito reviewed (changes requested):
The app doesn't work with the latest changes
![screenshot_2021-07-28_15-07-25](https://user-images.githubusercontent.com/43217/127402533-aa55f6cd-1a87-4e5b-a4d3-9e002ce67fd9.png)

Please make sure it does before submitting the PR :)

As for what's wrong, I did not debug the issue, but I guess you have to ditch your updater and instead trigger the update from this file somehow https://github.com/gnosis/cowswap/pull/953/files#diff-af06422a6866f6f150604ce5491f449819b0882344b8da8b2387f6288fe74949L1

I assume it's loading the url param too soon.
You can sync with Ramiro who coded that part, and as far I can tell, was working fine.

ramirotw commented:
> ![screenshot_2021-07-28_15-07-25](https://user-images.githubusercontent.com/43217/127402533-aa55f6cd-1a87-4e5b-a4d3-9e002ce67fd9.png)
> 
> Please make sure it does before submitting the PR :)
> 
> As for what's wrong, I did not debug the issue, but I guess you have to ditch your updater and instead trigger the update from this file somehow https://github.com/gnosis/cowswap/pull/953/files#diff-af06422a6866f6f150604ce5491f449819b0882344b8da8b2387f6288fe74949L1
> 
> I assume it's loading the url param too soon.
> You can sync with Ramiro who coded that part, and as far I can tell, was working fine.

The problem is that the `Updaters` component is rendered above in the tree than the `HashRouter`, so the location's context is not defined yet. A solution could be to move the Updaters below the HashRouter, or to instantiate the ReferralLinkUpdater as @alfetopito mentioned in the App file (btw, you should also remove the `useParseReferralQueryParam` reference in the `src/custom/pages/App/index.tsx` file which I left just for testing).

ramirotw commented:
@stackseeder Please fix the build error:

> src/custom/pages/App/index.tsx
> Line 1:17:  'useEffect' is defined but never used  @typescript-eslint/no-unused-vars

alfetopito reviewed (commented):
It all looks good now, but doesn't seem to be working.

I'm trying to debug this because I'm not sure whether this change has no effect of the URL loader is not working at all.

I followed the steps defined in the test section, but the `affiliateLink` on local storage is never updated
![screenshot_2021-07-29_11-59-09](https://user-images.githubusercontent.com/43217/127550468-2d06df76-83cc-4a1e-bc01-560029ad6ff6.png)

I've accessed it with a random `referral` https://pr1025--gpswapui.review.gnosisdev.com/#/swap?referral=adsadsa

Still empty on local storage.

Maybe is my setup. Can you try it out as well?

stackseeder commented:
@alfetopito sorry, the original PR was doing wrong action, so the original description was also wrong.
I've updated it.
I've also changed some names. It was storing referral link, but now storing referrer address, so I've changed the variable name from 'affiliateLink' to 'referrerAddress'.

alfetopito reviewed (approved):
Awesome!
